blockstore/splitstore/splitstore.go

```diff
@@ -1522,24 +1582,31 @@ func (s *SplitStore) purge(cids []cid.Cid, markSet MarkSet) error {
 		func(cids []cid.Cid) error {
 			deadCids := deadCids[:0]
 
-		again:
-			if err := s.checkClosing(); err != nil {
-				return err
-			}
+			for {
+				if err := s.checkClosing(); err != nil {
+					return err
+				}
 
-			s.txnLk.Lock()
-			if len(s.txnRefs) > 0 {
+				s.txnLk.Lock()
+				if len(s.txnRefs) == 0 {
+					// keep the lock when we hit the desired condition.
+					break
+				}
 				s.txnLk.Unlock()
 
+				// Doesn't really matter but...
+				//
+				// It's slightly strange that we're effectively looping _twice_.
+				// Maybe break protectTxnRefs into two functions? One for the loop,
+				// and one to do the actual thing?
```

Comment:
well, there is also logging involved, and I wanted to see those timings and numbers in the logs.